src/Logging.chpl

```diff
@@ -113,82 +102,60 @@ module Logging {
-                                            msg, "CRITICAL"));
-            } catch (e: Error) {
-                writeln(generateErrorMsg(moduleName, routineName, lineNumber, e));
-            }
+
+        // Emit `msg`, depending on `level`.
+        proc report(moduleName: string, routineName: string, lineNumber: int,
+                    level: LogLevel, msg:string...)
+        {
+            if level < this.level then return;
+
+            emit((...generateLogMessage(moduleName, routineName, lineNumber,
+                                        level, (...msg))));
         }
+
+        inline proc debug(moduleName, routineName, lineNumber, msg...) do
+          report(moduleName, routineName, lineNumber, LogLevel.DEBUG, (...msg));
+
+        inline proc info(moduleName, routineName, lineNumber, msg...) do
+          report(moduleName, routineName, lineNumber, LogLevel.INFO, (...msg));
+
+        inline proc warn(moduleName, routineName, lineNumber, msg...) do
+          report(moduleName, routineName, lineNumber, LogLevel.WARN, (...msg));
+
+        inline proc error(moduleName, routineName, lineNumber, msg...) do
+          report(moduleName, routineName, lineNumber, LogLevel.ERROR, (...msg));
         
-        proc generateErrorMsg(moduleName: string, routineName, lineNumber, 
-                           error) throws {
-            return "Error in logging message for %s %s %i: %?".format(
-                    moduleName, routineName, lineNumber, error.message());                
-        }
+        inline proc critical(moduleName, routineName, lineNumber, msg...) do
+          report(moduleName, routineName, lineNumber, LogLevel.CRITICAL, (...msg));
         
-        proc generateLogMessage(moduleName: string, routineName, lineNumber, 
-                           msg, level: string) throws {
-            var lineStr: string = if lineNumber != 0 then "Line %i ".format(lineNumber) else "";
-             if printDate {
-                 return "%s [%s] %s %s%s [Chapel] %s".format(
-                 generateDateTimeString(), moduleName,routineName,lineStr, 
-                                     level,msg);
-             } else {
-                 return "[%s] %s %s%s [Chapel] %s".format(moduleName, 
-                 routineName,lineStr,level,msg);            
-             }
+        proc generateLogMessage(moduleName: string, routineName: string, lineNumber: int,
+                                level: LogLevel, msg: string...) {
+            var lineStr = if lineNumber != 0 then "Line " + lineNumber:string + " " else "";
+            var dateStr = if printDate then generateDateTimeString(" ") else "";
+            return (dateStr, "[", moduleName, "] ", routineName, " ", lineStr, level:string,
+                    " [Chapel ] ", (...msg));
         }
          
-        proc generateDateTimeString() throws {
+        proc generateDateTimeString(tail = "") {
```

Comment:
What is the purpose of the `tail` argument and why is it only needed in the `catch` block?

Reply:
`tail` allows this helper to include a trailing space in the string it produces.

It is only needed in the catch block because I forgot to include it in the try block. I will fix this.

Reply:
Fixed as of c12932834

Reply:
Resolved